### PR TITLE
[SPARK-25565][BUILD] Add scalastyle rule to check add Locale.ROOT to .toLowerCase and .toUpperCase for internal calls

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR adds a rule to force `.toLowerCase(Locale.ROOT)` or `toUpperCase(Locale.ROOT)`.

It produces an error as below:

```
[error]       Are you sure that you want to use toUpperCase or toLowerCase without the root locale? In most cases, you
[error]       should use toUpperCase(Locale.ROOT) or toLowerCase(Locale.ROOT) instead.
[error]       If you must use toUpperCase or toLowerCase without the root locale, wrap the code block with
[error]       // scalastyle:off caselocale
[error]       .toUpperCase
[error]       .toLowerCase
[error]       // scalastyle:on caselocale
```

This PR excludes the cases above for SQL code path for external calls like table name, column name and etc.

For test suites, or when it's clear there's no locale problem like Turkish locale problem, it uses `Locale.ROOT`.

One minor problem is, `UTF8String` has both methods, `toLowerCase` and `toUpperCase`, and the new rule detects them as well. They are ignored.

## How was this patch tested?

Manually tested, and Jenkins tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -975,7 +975,7 @@ object SQLConf {
         "checkpoint location.")
       .stringConf
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: How about?
```scala
.transform(_.toLowerCase(Locale.ROOT))
.checkValue(
  str => Set("min", "max").contains(str)
```

**Review by srowen — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: not quite relevant with what the PR targets but okay.

**Comment by wangyum** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
rebased to reflect the latest change.

**Comment by HyukjinKwon** [comment]
just rebased to run the test to reflect the latest changes.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by HyukjinKwon** [comment]
Thanks for reviewing this @srowen and @wangyum

**Pull request closed without merging**